### DISCUSSION TITLE
Major Update #1: Gradio 4 Engine

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/stable-diffusion-webui-forge#853] [Announcements]
# New Gradio

![image](https://github.com/user-attachments/assets/0a60c7a3-bdbe-4ff3-a816-9aeb91a0692a)
![image](https://github.com/user-attachments/assets/ef7b2f7b-d56a-4857-8dca-e843aa1e78c5)

Gradio 4 finally landed. 

The updated version is Gradio 4.39.0, the lastest version at 2024 July 26.

We finally have this:

![1](https://github.com/user-attachments/assets/442773b0-33f0-4699-8609-96a49b717952)

Thanks to the very open HTML envs provided by Gradio 4, it is finally possible to control the website with nearly arbitary HTML/JS loaders. We worked with ChatGPT to created the **Forge Canvas** based on Gradio 4.

Keep this note:

**Left Mouse Button**: drag and drop image as a file, or draw on the canvas.

**Right Mouse Button**: drag the image.

**Mouse Wheel**: zoom in or out.

Note that to drag the image, you use **Right Mouse Button**! Not the left button! 

The left mouse button drag is for dragging the image as a file to other websites or even outside browser.

Basics:
![2](https://github.com/user-attachments/assets/b62efc80-f961-4618-941b-18b6a726c137)

Change View Size:
![3](https://github.com/user-attachments/assets/893c84dc-82d5-4251-b1c0-b2eea18f160d)

Perfect Transparency Support:
![4](https://github.com/user-attachments/assets/d7f261f8-9340-4e3c-a3a8-607039616a64)

Zoom-aware Brush Size: 
![5](https://github.com/user-attachments/assets/d6c605fe-c914-4693-a547-8a553ade9ddb)

Brush Width, Alpha, Softness:
![pen](https://github.com/user-attachments/assets/ee0f57df-b446-47d7-923e-6db1bcd5d639)
(128 level Wacom pen pressure support)
![6](https://github.com/user-attachments/assets/99d53717-5661-4bef-ada4-9a343123831f)

Extreme Detail Modification:
![7](https://github.com/user-attachments/assets/5324f4f5-ad74-4342-a9e9-85fff01a3ede)

Full Screen in One Click:
![8](https://github.com/user-attachments/assets/863a7f7d-74e7-40ee-8cb2-5406669525e4)

Mask Mode for Inpaint (now you never worry about brush color being invisible when it is similar to image color):
![9](https://github.com/user-attachments/assets/1644f03c-f402-4f74-a7c7-5572725efe4a)

Photoshop Clipboard Format Support:
![i](https://github.com/user-attachments/assets/0906ba16-af15-4003-9df5-97a49b9f96e1)

# Compatibility

Since Gradio 4 is very different, some extensions may be broken.

However, in my last test, about 70% extensions still work, including adetailer, prompt expansion, tagger, etc.. 

So maybe the problem is not that difficult.

If broken, most extensions only need several lines of codes to fix.

**Comment by lllyasviel** [comment]
see also https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/854

**Comment by themaxaboy** [comment]
@lllyasviel I've read your **Forge Status** about Gradio 4 UIs. Currently it's Normal status. I think I found a bug about using Batch with Upload or Path not working. Tested in img2img and extra tab.

**Comment by kusobabaa29** [comment]
Thank you for updating to gradio 4, it feels really comfortable to use for img2img, inpaint, etc. You're so amazing to make all these useful tools and make it available for everyone!

**Comment by aimarxjg** [comment]
感谢敏神，哇，小铅笔图标坏了1年多了！！！！终于修好了😫

**Comment by CCpt5** [comment]
Love that you're back working on Forge lllyasviel !!

Curious if anyone else is having an issue maximizing the live preview during inference w/ a double mouse click? The image will maximize fine w/ a double click if nothing is processing, but during inference a double click on the generating images doesn't open the live preview as expected.

Figure this may be a Gradio4 thing - but I'll try a clean install to see if it persists - if so I'll make a bug report......minor thing.

**Comment by Koxae** [comment]
The ADetailer is not working

**Comment by neldot70** [comment]
Hello. This latest update is a disaster for me. WEBUI Forge doesn't work anymore.

When launching it says "Stable diffusion model failed to load" and if I try to generate an image, every model I use it returns this error:
TypeError: 'NoneType' object is not iterable

Please, anyone can help? Can I downgrade to something that works? Thanks.

**Reply by Koxae** [reply]
Remove all extensions. One of them is incompatible with the new version. I had the same problem. Half of the extensions I use are incompatible with the new version of Forge

**Reply by neldot70** [reply]
Thanks for your help, Koxae.
I have just a basic understanding of the program, but I think I just installed several Loras and Checkpoints, but not other extensions. 

Where should I check and remove the extensions that you mention? Thanks in advance.

**Reply by Koxae** [reply]
Installed extensions are located at "Path to forge\webui\extensions". If this directory is empty, then the problem is somewhere else. Sorry if my advice does not help.

**Comment by ZafariJ0e** [comment]
after the latest update the loras no longer work , Does anyone have that problem?

**Comment by vhanla** [comment]
Gradio updated to 4.40.0 and index.js throws this error:
`Could not get api details: Request constructor: http://127.0.0.1:7860http://127.0.0.1:7860/info is not a valid URL.`
Tried with previous versions and --skip-install, and found that 4.36.1 doesn't fail on that part. I noticed the UI became too slow.
BTW: I'm using Python 3.11 on Windows 10 and 11 and disabled extensions.

**Reply by fgtm2023** [reply]
> Gradio updated to 4.40.0 and index.js throws this error: `Could not get api details: Request constructor: http://127.0.0.1:7860http://127.0.0.1:7860/info is not a valid URL.` Tried with previous versions and --skip-install, and found that 4.36.1 doesn't fail on that part. I noticed the UI became too slow. BTW: I'm using Python 3.11 on Windows 10 and 11 and disabled extensions.

thanks a lot for pointing that gradio is the problem, i solved my problem here: https://github.com/lllyasviel/stable-diffusion-webui-forge/issues/1562 by this: 
1- replaced gradio==4.40.0 to gradio==4.42.0 in requirements_versions.txt
2- pip install -r requirements_versions.txt --upgrade

**Reply by vhanla** [reply]
Actually, it still shows that error on 4.42.x and newer, I have tried 4.44.0 and it seems that gradio-client-1.4.0.tgz which contains the dist js scripts, doesn't make sure to check more than 2 colons in the URL in index.js file in this part:
```js
function resolve_root(base_url, root_path, prioritize_base) {
  if (root_path.startsWith("http://") || root_path.startsWith("https://")) {
    return prioritize_base ? base_url : root_path;
  }
  return base_url + root_path; //<---- this part is not error safe
}
``` 
I just used 7-zip to open index.js in dist directory within the archive, and edited to return only root_path instead of that concatenation. 
However, it has to be done after installing gradio with pip, due caching, and uninstalling or upgrading will install gradio-client-1.4.0.tgz in site-packages/gradio/_frontend_code/client within the python's lib path.

I suppose return base_url + root_path should validate a correct URI, because http://domain:port should allow two colons maximum, but in this case, it has 4 colons, which obviously is not valid. I hope they'll fix. However, I really don't know why this fails in the first place, as it seems other users do not face this issue.

**Comment by E2GO** [comment]
@lllyasviel Nice additions for inpaint process! Appreciated! I have some suggestions for the Inpainting Process (_because working a lot with inpainting and as a tablet user_):

1. Add an option to enable or disable mouse scroll on the inpaint window. Reason: When the mouse cursor is over the inpaint window and you try to scroll the Forge, you might accidentally zoom in or out of the inpaint window.
2. Separate the controls for resetting the mask editing and resetting the view, or add an enable/disable option for this feature. Now - clearing mask also resets window.
3. Introduce shortcuts for resizing the brush size, such as the [ ] keys, similar to Photoshop. This would be particularly useful when using a tablet.
4. Allow erasing already painted mask by holding down the ALT key.
5. Enable the ability to continue drawing the mask even when the cursor moves outside the canvas, preventing interruptions when the mouse goes out of the image.

**Comment by hpx502766238** [comment]
#[Image to Image, Resize and Fill](https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/12789),
#https://github.com/AUTOMATIC1111/stable-diffusion-webui/issues/10504
Even with the new Forge Canvas,this problem still exist:Cannot expand the canvas in selected directions,It always expands picture from center.

**Comment by photodarren** [comment]
I have a Mac and when I right click I get a menu. I can't move my image around. And if I maximise, I can't minimise. How is that supposed to be done? Esc key doesn't work. THere's no button to click on that I can find. How do I minimise, and how do I move my image with my mouse?

**Comment by xuguozhi** [comment]
How about sd-webui-controlnet extention?

**Comment by kevin20210917** [comment]
I placed the Forge Flux program on a UNC shared path. The program starts and runs normally, but it seems that Forge Canvas is not loading properly (it works fine without the Forge Canvas). How to solve it?

![Uploading _20250110154810.png…]()
